mypy/plugins/dataclasses.py

```diff
@@ -0,0 +1,295 @@
+            # even though it looks like a class-level assignment.
+            node.is_initialized_in_class = False
+
+            has_field_call, field_args = _collect_field_args(stmt.rvalue)
+
+            try:
+                # parse_bool returns an optional bool, so we corece it
+                # to a bool here in order to appease the type checker.
+                is_in_init = bool(ctx.api.parse_bool(field_args['init']))
+            except KeyError:
+                is_in_init = True
+
+            has_default = False
+            # Ensure that something like x: int = field() is rejected
+            # after an attribute with a default.
+            if has_field_call:
+                has_default = 'default' in field_args or 'default_factory' in field_args
+
+            # All other assignments are already type checked.
+            elif not isinstance(stmt.rvalue, TempNode):
+                has_default = True
+
+            known_attrs.add(lhs.name)
+            attrs.append(DataclassAttribute(
+                name=lhs.name,
+                is_in_init=is_in_init,
+                has_default=has_default,
+                line=stmt.line,
+                column=stmt.column,
+            ))
+
+        # Next, collect attributes belonging to any class in the MRO
+        # as long as those attributes weren't already collected.  This
+        # makes it possible to overwrite attributes in subclasses.
+        super_attrs = []
+        for info in cls.info.mro[1:-1]:
+            if 'dataclass' not in info.metadata:
+                continue
+
+            for name, data in info.metadata['dataclass']['attributes'].items():
+                if name not in known_attrs:
+                    attr = DataclassAttribute.deserialize(data)
+                    known_attrs.add(name)
+                    super_attrs.append(attr)
+
+        all_attrs = super_attrs + attrs
+
+        # Ensure that arguments without a default don't follow
+        # arguments that have a default.
+        found_default = False
+        for attr in all_attrs:
+            # If we find any attribute that is_in_init but that
+            # doesn't have a default after one that does have one,
+            # then that's an error.
+            if found_default and attr.is_in_init and not attr.has_default:
+                ctx.api.fail(
+                    'Attributes without a default cannot follow attributes with one',
+                    Context(line=attr.line, column=attr.column),
+                )
+
+            found_default = found_default or attr.has_default
+
+        return all_attrs
+
+    def _freeze(self, attributes: List[DataclassAttribute]) -> None:
+        """Converts all attributes to @property methods in order to
+        emulate frozen classes.
+        """
+        info = self._ctx.cls.info
+        for attr in attributes:
+            try:
+                node = info.names[attr.name].node
```

Comment:
I think this will look cleaner if refactored via `info.names.get(attr.name)` which cannot raise.